adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/MqttContext.java

```diff
@@ -34,92 +38,136 @@
  */
 public final class MqttContext extends MapBasedTelemetryExecutionContext {
 
-    private MqttPublishMessage message;
-    private MqttEndpoint deviceEndpoint;
-    private Device authenticatedDevice;
-    private ResourceIdentifier topic;
+    private final MqttPublishMessage message;
+    private final MqttEndpoint deviceEndpoint;
+    private final ResourceIdentifier topic;
+    private final TenantObject tenantObject;
+    private final String authId;
+    private final PropertyBag propertyBag;
+    private final MetricsTags.EndpointType endpoint;
+
     private String contentType;
     private Sample timer;
-    private MetricsTags.EndpointType endpoint;
-    private PropertyBag propertyBag;
 
-    private MqttContext(final Device authenticatedDevice) {
+    private MqttContext(
+            final MqttPublishMessage message,
+            final MqttEndpoint deviceEndpoint,
+            final ResourceIdentifier topic,
+            final TenantObject tenantObject,
+            final String authId,
+            final Device authenticatedDevice,
+            final PropertyBag propertyBag) {
         super(authenticatedDevice);
-    }
-
-    @Override
-    public QoS getRequestedQos() {
-
-        switch (message.qosLevel()) {
-            case AT_LEAST_ONCE:
-                return QoS.AT_LEAST_ONCE;
-            case AT_MOST_ONCE:
-                return QoS.AT_MOST_ONCE;
-            default:
-                return QoS.UNKNOWN;
-        }
+        this.message = Objects.requireNonNull(message);
+        this.deviceEndpoint = Objects.requireNonNull(deviceEndpoint);
+        this.topic = Objects.requireNonNull(topic);
+        this.tenantObject = Objects.requireNonNull(tenantObject);
+        this.authId = Objects.requireNonNull(authId);
+        this.propertyBag = propertyBag;
+        this.endpoint = MetricsTags.EndpointType.fromString(topic.getEndpoint());
     }
 
     /**
      * Creates a new context for a published message.
      *
-     * @param publishedMessage The MQTT message to process.
-     * @param deviceEndpoint The endpoint representing the device
-     *                       that has published the message.
-     * @return The context.
-     * @throws NullPointerException if message or endpoint are {@code null}.
+     * @param publishedMessage The published MQTT message.
+     * @param deviceEndpoint The endpoint representing the device that has published the message.
+     * @param tenantAndAuthIdExtractor The extractor to determine tenant and authentication identifier from the CONNECT
```

Comment:
Since the authentication identifier isn't determined from the device, I'd rather use:
````
The extractor to determine tenant and authentication identifier from the
authentication data used during connection establishment or from the topic of the PUBLISH message in case
of an unauthenticated connection.
````

Reply:
In case of an unauthenticated connection there is no *authentication identifier* but only a *device identifier* ...

Reply:
Yes, ok, adapted it accordingly.